astroid/cache.py

```diff
@@ -0,0 +1,104 @@
+""" Caching utilities used in various places."""
+
+import typing
+from collections import OrderedDict
+from weakref import WeakSet
+
+import wrapt
+
+LRU_CACHE_CAPACITY = 128
+
+K = typing.TypeVar("K")
+V = typing.TypeVar("V")
+
+
+class LRUCache(typing.Generic[K, V]):
+    """An LRU cache that keeps track of its instances."""
+
+    instances: "WeakSet[LRUCache[typing.Any, typing.Any]]" = WeakSet()
+
+    def __init__(self) -> None:
+        self.cache: typing.OrderedDict[K, V] = OrderedDict()
+        LRUCache.instances.add(self)
+
+    def __setitem__(self, key: K, value: V) -> None:
+        self.cache[key] = value
+
+        if len(self.cache) > LRU_CACHE_CAPACITY:
+            self.cache.popitem(last=False)
+
+    def __getitem__(self, key: K) -> V:
+        if key in self.cache:
+            self.cache.move_to_end(key)
+
+        return self.cache[key]
+
+    def __contains__(self, key: K) -> bool:
+        return key in self.cache
+
+    def clear(self) -> None:
+        self.cache.clear()
+
+    @classmethod
+    def clear_all(cls) -> None:
+        """Clears all LRUCache instances."""
+        for cache in cls.instances:
+            cache.clear()
+
+
+def lru_cache_astroid(arg=None):
+    """A decorator to cache the results of a function. Similar to
+    functools.lru_cache but uses astroid.cache.LRUCache as its internal cache.
+    """
+    cache = LRUCache()
+
+    @wrapt.decorator
+    def decorator(func, instance, args, kwargs):
+        key = (instance,) + args
+
+        for kv in kwargs:
+            key += kv
```

Comment:
Based on this wouldn't the `key` for a `LRUCache` always be a `Tuple[Any]`?

Reply:
You are right, I've typed it as `typing.Tuple[typing.Any, ...]`.